Add aten::flatten NNAPI converter

axitkhurana opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61025
* #61024
* #61023
* #61022
* #61021
* #59529
* #59434
* #59365
* #59364
* #58543
* **#60885**
* #58541
* #58540
* #58539

Summary:

Add support for aten::flatten op in the NNAPI model converter.
Variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

Commit:
Add flatten WIP

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "Add flatten WIP"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

Commit:
Update on "Add aten::flatten NNAPI converter"


Summary:

Add support for aten::div op in the NNAPI model converter. Startup time
variable size support isn't supported as shapes go as inputs to NNAPI op

Runtime variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

[ghstack-poisoned]

Commit:
Update on "Add aten::flatten NNAPI converter"


Summary:

Add support for aten::div op in the NNAPI model converter. Startup time
variable size support isn't supported as shapes go as inputs to NNAPI op

Runtime variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

Commit:
Update on "Add aten::flatten NNAPI converter"


Summary:

Add support for aten::div op in the NNAPI model converter. Startup time
variable size support isn't supported as shapes go as inputs to NNAPI op

Runtime variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

Commit:
Update on "Add aten::flatten NNAPI converter"


Summary:

Add support for aten::div op in the NNAPI model converter. Startup time
variable size support isn't supported as shapes go as inputs to NNAPI op

Runtime variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

Commit:
Update on "Add aten::flatten NNAPI converter"


Summary:

Add support for aten::div op in the NNAPI model converter. Startup time
variable size support isn't supported as shapes go as inputs to NNAPI op

Runtime variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -918,6 +921,53 @@ def add_reshape(self, node):
+
+        in_id, in_oper = self.get_tensor_operand_by_jitval_fixed_size(node.inputsAt(0))
+
+        start_ctype, start_dim = self.get_constant_value(node.inputsAt(1))
+        end_ctype, end_dim = self.get_constant_value(node.inputsAt(2))
```
Force these to be ints, like you did for softmax dim?

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -918,6 +921,53 @@ def add_reshape(self, node):
+        #     for dim in out_shape
+        # )
+
+        out_oper = in_oper._replace(shape=out_shape, dim_order=DimOrder.PRESUMED_CONTIGUOUS)
```
Order is already contiguous.

Commit:
Update on "Add aten::flatten NNAPI converter"


Summary:

Add support for aten::flatten op in the NNAPI model converter.
Variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

[ghstack-poisoned]

Commit:
Update on "Add aten::flatten NNAPI converter"


Summary:

Add support for aten::flatten op in the NNAPI model converter.
Variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

[ghstack-poisoned]

Commit:
Update on "Add aten::flatten NNAPI converter"


Summary:

Add support for aten::flatten op in the NNAPI model converter.
Variable size support to supported soon

Test Plan:

pytest test/test_nnapi.py::TestNNAPI::test_flatten

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29451725](https://our.internmc.facebook.com/intern/diff/D29451725)

[ghstack-poisoned]

axitkhurana commented:
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29451725).
